### PR TITLE
Experiments on top of PR #3643 (pybind11_select_caster)

### PR DESCRIPTION
**Pull request by rwgk**
## Description
WIP on top of #3643.

## Suggested changelog entry:

```rst

```

### CODE REVIEW

File: docs/advanced/cast/custom.rst [hunk 1]
```diff
@@ -31,63 +31,83 @@ The following Python snippet demonstrates the intended usage from the Python sid
...
+          template <> struct type_caster<inty> : inty_type_caster {};
+      }} // namespace pybind11::detail
+
+  .. warning::
+
+      When using this method, it's important to declare them consistently
+      in every compilation unit of the Python extension module. Otherwise,
+      undefined behavior can ensue.
+
+* The preferred method is to *declare* a function named
```
Comment (amauryfa) [inline docs/advanced/cast/custom.rst:104]: "When you own the namespace where the type is defined, the preferred method..."

**Comment by rwgk** [comment]
The force push was purely for `git rebase master`.

**Comment by laramiel** [comment]
I have reworked this on top of HEAD with #3862

**Comment by rwgk** [comment]
Work continued under #3862.

**Pull request closed without merging**